➕ 1단계 - 문자열 덧셈 계산기

boradol opened the pull request:
안녕하세요. 리뷰어님.
문자열 덧셈 계산기 제출합니다!!

고민한 부분해서 작성한 부분은
1. `Number` Wrapper Class를 하였습니다. 음수가 아닌 수여서 `Number`라는 클래스 명이 고민 되었습니다.
2. 정적 팩토리 메서드를 이용하여 Number를 생성하도록 하였습니다.
3. Number에서도 `null`이 들어오면 0을 반환하고  `NumbersParserUtils.parse` 에서도 한번 더 null 체크를 하였습니다. 뭔가 두번 하는 느낌인데 괜찮을까요?
4. 'Numbers'라는 일급객체를 사용해보았습니다. 이 객체가 총 합을 계산하도록 책임을 주었습니다.
5. 함수 명을 지을 때 객체가 수행을 하는 것에 초점을 맞춰 메서드명을 작성하려고 하였는데요 어려웠습니다.
6. 구분자에 관련해서 전략 패턴을 사용하였습니다. `ServiceLoader.load(~Strategy.class)` 를 사용해서 전략 구현체들을 받아오고 싶었는데 테스트에서 작동하지 않았습니다. 할수없이 List.of를 사용해서 불러왔는데요. 그럼 `NumbersParserUtils`에서 OCP를 위배하는 것 같아서 고민이 되었습니다.
7. `StringCalculatorTest`의 예시 코드 기반으로 하면 int끼리 비교하는데요. Number로 즉 객체끼리 비교하는 것이 나을까요? 뭔가 getValue()하게 되는 것 같아서 수정할까 말까 고민하였습니다.


고민한 부분에 대해 이야기 해주실 부분이 있으시면 조언 부탁드립니다>.<
많이 부족한 실력입니다. 정말 많이 배워가고 싶습니다.
그리고 앞으로 미션 진행하는데 잘 부탁드립니다 🙏🏻

boradol committed:
docs: 요구사항 정리

boradol committed:
test(stringcalculator): 문자열 계산기 테스트 코드 작성

boradol committed:
test(Number): 입력받은 문자열이 음수나 숫자가 아닌 문자를 입력받을때 예외 발생하는 테스트 코드 작성

boradol committed:
feat(Number): 입력받은 문자열이 음수나 숫자가 아닌 문자 일때 Number를 생성하는 기능 추가

1. 숫자가 아닌 문자열일 경우 예외 발생
2. 음수일 경우 예외 발생

boradol committed:
test(Number): 0을 반환하는 경우의 테스트 작성

boradol committed:
feature(Number): 0을 반환하는 기능 추가

1. Number는 무조건 정적 팩토리 메서드를 이용해서만 생성됨.
2. null, empty, blank 모두 0을 반환하는 기능.

boradol committed:
test(Number): 유효한 숫자에 대한 테스트 코드 작성

boradol committed:
feat(Number): Number에 대한 값을 반환하는 함수 추가

boradol committed:
test(Number): 두 수를 더한 결과를 반환하는 테스트 작성

boradol committed:
feat(Number): 덧셈 기능 추가

boradol committed:
test(Numbers): 숫자들 배열의 합 테스트 작성

boradol committed:
feat(Numbers): 일급객체 클래스 생성 및 숫자들의 합을 계산 기능 추가

boradol committed:
feat(NumberSplitStrategy): 전략 패턴 -> 컴마와 콜론에 대한 구분자 기능 구현

1. Add BasicSplitStrategy

boradol committed:
feat(NumberSplitStrategy): 커스텀 구분자 처리에 대한 전략 추가

1. Add CustomSplitStrategy

boradol committed:
feat(NumbersParserUtils): 구분자가 포함된 전략에 따라 Numbers 파싱하는 유틸 추가

boradol committed:
feat(StringCalculator): 문자열 계산기 add 함수 구현

wlroh commented on README.md:
```diff
@@ -16,3 +16,29 @@ docker compose -p kitchenpos up -d
+  - [ ] 컴마와 콜론에 대한 처리
+  - [ ] custom 구분자에 대한 처리
+- [ ] 합을 반환하는 기능
+  - [ ] 숫자가 두개 이상일 경우를 생각해서 숫자 리스트를 일급 객체로 반환
```
요구사항과 요구사항으로 부터 기능 문서화 습관 👍 💯

wlroh commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @DisplayName(value = "빈 문자열 또는 null 값을 입력할 경우 0을 반환해야 한다.")
+    @ParameterizedTest
+    @NullAndEmptySource
```
`@ParameterizedTest` 와 `@NullAndEmptySource` 사용 👍 
여러케이스를 하나의 테스트코드로 검증할 수 있죠!

wlroh commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,63 @@
+    @DisplayName(value = "숫자 하나를 문자열로 입력할 경우 해당 숫자를 반환한다.")
+    @ParameterizedTest
+    @ValueSource(strings = {"1"})
+    void oneNumber(final String text) {
+        assertThat(calculator.add(text)).isSameAs(Integer.parseInt(text));
+    }
```
검증하는 부분을 `Integer.parseInt(text)` 로 사용하신 의도가 있을까요? 🤔 
어떤 것을 검증하시려는 것일까요?

boradol replied:
이 부분 같은 경우에는 강의 교안에 올라와 있는 그대로 복사해서 가져와서 그렇게 된것 같습니다.
굳이 파싱하지 않고 바로 int형으로 했어도 되었는데.... 
수정하도록 하겠습니다. :)

boradol replied:
https://github.com/next-step/ddd-legacy/pull/590/commits/6bab14e113c262b59d744c71fb8440b935d30951

wlroh replied:
보라님도 아실 것 같은데 검증에 대해서는 하드코딩이 좋다고 생각해요 😄 
관련해서 좋은 아티클이 있어서 공유드릴게요
https://jojoldu.tistory.com/615

boradol replied:
감사합니다. 주신 좋은 글 읽어보도록 하겠습니다.

wlroh commented on src/main/java/calculator/Number.java:
```diff
@@ -0,0 +1,30 @@
+
+    private static void validateNumber(final int value) {
+        if (value < ZERO) {
+            throw new IllegalArgumentException(NEGATIVE_NUMBER_EXCEPTION);
```
에외와 상세메시지 👍 
다만, 상세메시지에 실패관련 정보도 같이 담아보는 것은 어떨까요?
추후 운영 중에 이슈가 발생하게 되었을 때 실패관련 정보까지 있으면 디버깅하는데 많은 도움이 될 수 있어요.

boradol replied:
오 그러게요. 이렇게 하면 어떤 값이 잘못 넘겨졌는지 볼 수 있겠네요.
실패 관련 정보인 value값도 같이 넘기게 수정하겠습니다 :)

boradol replied:
https://github.com/next-step/ddd-legacy/pull/590/commits/3553f8dd09cfa122688d133d36c829d576587b49

wlroh commented on src/main/java/calculator/Number.java:
```diff
@@ -27,4 +34,17 @@ private static int parseInt(final String value) {
         }
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Number number = (Number) o;
+        return value == number.value;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(value);
+    }
```
`equals()`, `hashCode()` 값객체로써 동등성비교하기 위한 재정의 👍 💯 
테스트코드 검증에서도 값객체로 바로 비교도 👍

wlroh commented on src/test/java/calculator/NumberTest.java:
```diff
@@ -46,4 +46,19 @@ void validNumber() {
+        Number result = firstOperand.plus(secondOperand);
+
+        // then
+        assertThat(result).isEqualTo(Number.from("7"));
+        assertThat(result.value()).isEqualTo(7);
```
> assertThat(result).isEqualTo(Number.from("7"));

`두 수를 더하여 덧셈의 결과인 숫자를 반환한다.` 에 대해서는 위의 검증 하나로도 충분하다고 생각이 드는데 어떻게 생각하세요?

저는 테스트코드 또한 유지보수해야하는 코드라고 생각해요.
그래서 불필요한? 과한? 검증은 오히려 생산성을 방해하는 요인으로 생각해요

boradol replied:
이건... 제가 Number로 검증해야할지 int로 검증해야할지 고민을 한 흔적인것 같습니다.
수정하도록 하겠습니다.!!

boradol replied:
https://github.com/next-step/ddd-legacy/pull/590/commits/6bab14e113c262b59d744c71fb8440b935d30951

wlroh commented on README.md:
```diff
@@ -36,7 +36,8 @@ docker compose -p kitchenpos up -d
   - [x] 정적 팩토리 메서드 사용하여 Number 생성.
   - [x] `empty`이나 `null`에 대해 0 반환 (더하여 blank 처리도...)
-  - [x] 숫자일 경우 숫자 한개만 반환 
+  - [x] 숫자일 경우 숫자 한개만 반환
+  - [x] 덧셈 기능 추가
```
꾸준히 요구사항 문서화 업데이트하는 습관 👍 👍

wlroh commented on src/main/java/calculator/strategy/BasicSplitStrategy.java:
```diff
@@ -0,0 +1,21 @@
+
+public class BasicSplitStrategy implements NumbersSplitStrategy {
+
+    private static final String DELIMITER_REGEX = "[,:]";
+    private static final Pattern MATCH_PATTERN = Pattern.compile("-?\\d+(?:[,:]-?\\d+)*");
```
정규식 관련 객체는 생성비용이 비싼데 재사용할 수 있도록 상수로 등록 👍 💯

wlroh commented on src/main/java/calculator/strategy/BasicSplitStrategy.java:
```diff
@@ -0,0 +1,21 @@
+    private static final String DELIMITER_REGEX = "[,:]";
+    private static final Pattern MATCH_PATTERN = Pattern.compile("-?\\d+(?:[,:]-?\\d+)*");
+
+    @Override
+    public Numbers extract(String input) {
+        return new Numbers(input.split(DELIMITER_REGEX));
+    }
+
+    @Override
+    public boolean isMatchPattern(String input) {
+        return MATCH_PATTERN.matcher(input).matches();
+    }
```
전략패턴 관련해서 테스트코드는 없는것 같아요. 😢

boradol replied:
전략패턴도 테스트 코드를 구현해야 하군요.
그러니까 모든 클래스에 관련해서 테스트코드를 짜야함을 간과하지 않겠습니닷!! 🫡

boradol replied:
https://github.com/next-step/ddd-legacy/pull/590/commits/829e4e68fcee2ae9518cfd227935cb7b9568fdbb

wlroh replied:
> 그러니까 `모든 클래스`에 관련해서 테스트코드를 짜야함을 간과하지 않겠습니닷!! 🫡

모든 클래스에 관해 테스트 코드를 작성할 필요는 없어요!
always, all 이라는 단어때문에 혹시나해서 말씀드려요.
테스트커버리지와 관련된 내용이기도 한데 저는 개인적으로 테스트커버리지는 100% 채우기 위해 비즈니스적으로 의미없는 getter 혹은 비즈니스적으로 의미없는 클래스에 대한 테스트코드까지 작성하는 것이면 오히려 유지보수측면에 악영향을 끼친다고 생각해요

즉, 비즈니스적으로 유의미한 로직에 대해서만 테스트코드를 작성하면 된다고 생각해요.

boradol replied:
> 비즈니스적으로 유의미한 로직에 대해서만 테스트코드를 작성하면 된다고 생각해요.

아 네네! 알겠습니다.
저도 멘토님의 말과 같은 의미로 말씀드리려고 한것인데... 제 단어선택이 모든 퍼블릭 인터페이스를 테스트 한다로 들리셨겠군요!
멘토님의 혹시나 하는 노파심으로 진심으로 해주신 코멘트같아서 더 좋습니다. 감사합니다. 🙇‍♀️

wlroh commented on src/main/java/calculator/NumbersParserUtils.java:
```diff
@@ -0,0 +1,26 @@
+                .findAny()
+                .map(strategy -> strategy.extract(input))
+                .orElse(ZERO_NUMBERS);
+    }
```
유틸성 객체 생성못하도록 private 접근제어자 👍 💯 
좀더 나아가면 생성자에서 exception 을 던져서 내부에서도 생성을 못하도록 막아줄수도 있어요.
해당 객체도 테스트코드가 없는데 확인부탁드려요.

boradol replied:
오 내부에서도 생성을 막아준다.
이 부분 미처 생각하지 못했던 부분인데 감사합니다:)

테스트코드 추가하도록 하겠습니다!!

boradol replied:
https://github.com/next-step/ddd-legacy/pull/590/commits/54c82190333ab56b4b1a79b072119d2d0e0a264e
https://github.com/next-step/ddd-legacy/pull/590/commits/c68987a0b988979ee0bf8ce770b0e337bbaeb4a4

wlroh commented on src/test/java/calculator/NumbersTest.java:
```diff
@@ -0,0 +1,21 @@
+
+class NumbersTest {
+
+    @DisplayName("문자열의 배열을 입력 받아 합의 결과를 반환한다.")
+    @Test
+    void sumTest() {
+        // given
+        Numbers numbers = new Numbers("1", "2", "3", "4");
+
+        // when
+        Number result = numbers.sum();
+
+        // then
+        Assertions.assertThat(result).isEqualTo(Number.from("10"));
+    }
```
DisplayName 을 다듬어보는 것은 어떤가요?
시나리오? 흐름을 나타내는 것으로 느껴져서요.
테스트 코드를 작성하는 이유와 역할은 어떤 것이라고 생각하세요? 🤔

boradol replied:
DisplayName 다듬어 수정해보겠습니다!
DisplayName 작성 이유와 역할도 생각해보았습니다.
- 가독성 향상
- 유지보수 향상
- 테스트 결과 이해
이런 것들을 생각하면서 작성해보도록 하겠습니다 :)

boradol replied:
https://github.com/next-step/ddd-legacy/pull/590/commits/842d76e62407b80156fe100ccecf6510eb05cb55

wlroh commented:
보라님 안녕하세요 😄 
미션 빠르게 해주셨네요 👍 

미션하면서 많은 고민거리가 있으셨군요. 
개발은 정답이 없고 대부분이 트레이드 오프 영역이라 생각해서 이런 토론들을 좋아해요. 

우선 정리해주신 고민들에 대해 제 의견 말씀드릴게요.

> Number Wrapper Class를 하였습니다. 음수가 아닌 수여서 Number라는 클래스 명이 고민 되었습니다.

> 함수 명을 지을 때 객체가 수행을 하는 것에 초점을 맞춰 메서드명을 작성하려고 하였는데요 어려웠습니다.

두개의 고민이 연관되어서 같이 묶었어요. 네이밍에 대해서는 저는 이해관계자분들만 모두 동의하고 해당 네이밍을 들었을 때 같은 것을 떠올릴수만 있다면 아무 문제가 없다고 생각해요. 또한, PR 리뷰를 하는 이유가 네이밍에 대해서도 서로의 의견을 주고받아서 모두가 동의하는 코드를 작성할 수 있다는 것이 아닐까해요

> 정적 팩토리 메서드를 이용하여 Number를 생성하도록 하였습니다.

정적 팩토리 메서드 👍 👍 생성자 접근제어자도 private 으로 구현해주신 것도 확인했어요 ㅎㅎ
코멘트가 많아서 리뷰달려다가 패스한 것이 있는데 정적 팩토리 메서드를 언급하셨으니.. 의견하나 더 드리면, 정적팩터리메서드를 활용하면 캐싱된 데이터를 제공할 수 있는 장점이 있죠. 이미 상수로 리턴함으로써 재사용하도록 구현하셨는데 이왕 구현하신거 input 값이 "0" 일 때도 상수 제공하면 어떨까 생각이 들었어요

> Number에서도 null이 들어오면 0을 반환하고 NumbersParserUtils.parse 에서도 한번 더 null 체크를 하였습니다. 뭔가 두번 하는 느낌인데 괜찮을까요?

네네 이부분도 트레이드오프 영역이여서 나쁘지는 않았어요. 더 꼼꼼하게 체크되는 것이니까요 ㅎㅎ
Wrapper Class 를 사용하는 장점 중 하나가 이미 검증이 완료되서 믿고 쓸 수 있다는 것이니까요.
다만, 저는 실무에서는 앞단쪽에서 null 에 대한 검증이 되면 뒤쪽은 검증은 안하는 편이에요.

> 'Numbers'라는 일급객체를 사용해보았습니다. 이 객체가 총 합을 계산하도록 책임을 주었습니다.

일급객체 👍 👍 앞으로도 미션진행하면서 일급객체 사용해보시면서 장단점을 느껴보셨으면 좋겠어요 

> 구분자에 관련해서 전략 패턴을 사용하였습니다. ServiceLoader.load(~Strategy.class) 를 사용해서 전략 구현체들을 받아오고 싶었는데 테스트에서 작동하지 않았습니다. 할수없이 List.of를 사용해서 불러왔는데요. 그럼 NumbersParserUtils에서 OCP를 위배하는 것 같아서 고민이 되었습니다.

저는 원칙을 무조건 지켜야만 한다고 생각하지않아요. 그렇다고 원칙을 무시하라는 의미는 아니구요 ㅎㅎ 원칙이 나오게 된 이유? 본질에 집중하시고 선택하시면 좋을 것 같아요.
OCP 의 경우 결국 변경에 자유롭고 싶다는 것인데, 현재 구현해주신 것으로 새로운 전략패턴이 추가되었다고 NumbersParserUtils 로직에 많은 영향을 주는 구조로 보이지 않아서 괜찮다고 생각해요.
다만, 매번 NumbersParserUtils 에 전략패턴을 추가해주는 작업이 번거롭거나 놓치는 휴먼에러가 걱정되시면 스프링 사용한다는 가정하에 빈으로 등록하거나 Enum 을 활용하는 방법도 있을 것 같네요

> StringCalculatorTest의 예시 코드 기반으로 하면 int끼리 비교하는데요. Number로 즉 객체끼리 비교하는 것이 나을까요? 뭔가 getValue()하게 되는 것 같아서 수정할까 말까 고민하였습니다.

StringCalculatorTest의 검증 부분을 int 로 할지, Number로 할지 고민이라는 것일까요? 🤔 
StringCalculatorTest 의 경우는 저는 int 로 비교하는게 좋을 것 같아요. 외부로부터 input 값을 받고 output 을 제공하는 객체이기 때문에 결과값이 Number 가 아닌 int 형이여야하지 않을까하네요
Number 로 리턴하면 해당 객체를 의존하는 쪽에서 Number 까지 같이 의존하게 되어서요

wlroh reviewed (changes requested):
보라님 안녕하세요 😄 
미션 고생하셨습니다 👍 
코멘트 달아놓았는데요, 확인부탁드려요!
궁금하신점 있으시면 PR, DM 편하게 연락주세요

boradol committed:
docs(README.md): 리뷰사항 목록 업데이트

boradol committed:
refactor(StringCalculatorTest): 테스트 코드에 과도한 구현 제거

1. NumberTest : 과한 검증 제거
2. StringCalculatorTest : 불필요한 Integer 변환 제거

boradol committed:
refactor(NumbersTest): DisplayName 변경 및 DisplayName 의미와 역할 정리하기

boradol committed:
refactor(Number): 실패 관련 메세지를 위한 exception 변경

1. IllegalNumberException 추가
2. Invalid 한 number 정보 추가

boradol committed:
refactor(NumbersParserUtils): private 생성자에 생성시 exception 발생하도록 추가

boradol committed:
test(NumbersParserUtilsTest): 파싱하는 함수의 테스트 코드 작성

1. null or empty 입력시 0 반환 테스트
2. 구분자포함한 문자열 들어왔을 시에
3. 유효하지 않은 구분자를 입력 할 시 에러 발생
4. IllegalDelimiterArgumentException 추가
5. ZeroNumbersSplitStrategy 전략 추가

boradol committed:
test(NumbersSplitStrategyTest): 숫자 분리하는 전략 패턴 테스트 작성

boradol committed:
refactor : 멤버변수와 정적변수 구분하기 위한 개행

boradol commented:
안녕하세요. 리뷰어님😃
모든 고민거리에 멋진 답변을 해주셔서 감사합니다.👍

> 개발은 정답이 없고 대부분이 트레이드 오프 영역이라 생각해서 이런 토론들을 좋아해요.

이렇게 좋은 말씀도 덧붙여 주셔서 정말 감사합니다!
전반적으로 제가 트레이드오프 상황에 대해 질문을 많이 드린 것 같아요.
아직 저는 개발적으로 이렇게 코드를 짜야함에 있어 근거가 부족하기에 고민이 되었던 것 같아요.
그래도 여러가지 가능성에 대해 이야기 해주시면서 리뷰어님의 개발적 철학도 엿볼 수 있어 좋았습니다.

추가로
제가 맨 처음에 PR메세지에 남긴 부분에서 추가 리뷰 주신 부분
>  input 값이 "0" 일 때도 상수 제공하면 어떨까 생각이 들었어요. 
- https://github.com/next-step/ddd-legacy/pull/590/commits/6940ef1f50672e31ab7dec35ee6a56adc98762a9
이 부분에 대해서 수정해 보았는데요. 리뷰어님이 말씀하신대로 제가 잘 이해하고 적용한건지 모르겠더라구요ㅠ
아래 커밋 부분 남깁니다. 확인 부탁드려용~!! 

저도 리뷰어님께 리뷰 받는동안에는 리뷰어님의 조언에 충실하고 싶습니다.
많이 배우고 싶습니다. 
그럼 계속해서 리뷰 잘 부탁드립니다. 🙏🏻

boradol committed:
refactor(Number) : input "0"일 때, 상수 제공

wlroh commented on src/test/java/calculator/NumbersTest.java:
```diff
@@ -6,7 +6,7 @@
 class NumbersTest {
 
-    @DisplayName("문자열의 배열을 입력 받아 합의 결과를 반환한다.")
+    @DisplayName("입력받은 숫자들의 합을 반환한다.")
```
변경해주신 내용도 좋지만 `입력받은 숫자들` 이라는 의미가 파라미터를 생각하게 되기도하고, Numbers 를 생성과정에 대한 내용도 같이 표현한 것으로 생각들었어요.

`해당 케이스에서 검증하는 것이 무엇일까?` 생각했을 때 저는 Numbers.sum() 에 대한 검증이라고 생각하고 Numbers 객체에서 sum이라는 행위는 Number 의 총합을 제공하는 것이 아닐까 생각이 들어요.

즉, Numbers 입장에서는 Number 가 입력받은 숫자인지, DB에 저장된 숫자인지 중요하지 않다고 생각해요.
Numbers 객체가 어떤 역할을 수행하는지로 보게되면 아래와 같이도 표현할 수 있지 않을까해요.

> Numbers 는 Number의 총합을 제공한다.

https://github.com/next-step/ddd-legacy/pull/590#discussion_r1587800188

코멘트를 드리게 된 계기는 테스트 코드에 대해 잘 작성하고 DisplayName 에 또한 잘 작성하다보면 자연스럽게 요구사항 명세서의 역할까지 해줄수있다고 봐요.
테스트코드를 작성함에 따라 보너스로 요구사항 명세서까지 할 수 있는거죠.

boradol replied:
좋은 리뷰였는데 지나칠뻔 했네요.

> 즉, Numbers 입장에서는 Number 가 입력받은 숫자인지, DB에 저장된 숫자인지 중요하지 않다고 생각해요.
Numbers 객체가 어떤 역할을 수행하는지로 보게되면 아래와 같이도 표현할 수 있지 않을까해요.

이 부분을 조금 주의깊게 읽었다면 2단계 요구사항 때 많은 도움이 되었을 것 같다라는 생각이 듭니다.

이부분도 3단계 미션에 같이 수정하겠습니다.

wlroh commented on src/main/java/calculator/Number.java:
```diff
@@ -31,15 +33,16 @@ public int value() {
         } catch (NumberFormatException e) {
-            throw new IllegalArgumentException(PARSING_INTEGER_EXCEPTION);
+            throw new IllegalNumberException(PARSING_INTEGER_EXCEPTION, value);
+
```
커스텀예외 👍 
커스텀 예외의 장점중 하나는 예외명을 통해 명시적으로 어떤 예외인지 체크가 가능하다는점과 그렇기 때문에 후처리 또한 정교하게 할 수 있다는 점이 있어요.
이왕 커스텀예외로 만드는 것이면 `음수 예외`와 `숫자가 아닌 문자열 예외` 2개를 만들고 각각의 커스텀예외에 Default 메시지를 관리하는 것도 좋지 않을까 생각이 들어요. 
혹은 표준예외를 이용하고 상세메시지와 실패 관련 정보를 제공할 수 도 있고요.

개인적인 의견이여서 꼭 수정하지 않으셔도돼요.

boradol replied:
3단계 미션때 함께 반영해보겠습니다.

> 이왕 커스텀예외로 만드는 것이면 음수 예외와 숫자가 아닌 문자열 예외 2개를 만들고 각각의 커스텀예외에 Default 메시지를 관리하는 것도 좋지 않을까 생각이 들어요.

이것으로 수정해 보겠습니다.

wlroh commented on src/main/java/calculator/NumbersParserUtils.java:
```diff
@@ -1,27 +1,28 @@
                 .findAny()
                 .map(strategy -> strategy.extract(input))
-                .orElse(ZERO_NUMBERS);
+                .orElseThrow(() -> new IllegalDelimiterArgumentException(INVALID_DELIMITER_MESSAGE, input));
```
여기도 커스텀예외인데 메시지를 생성자로 받지말고 커스텀예외가 기본메시지를 상수로 관리할 수 있을 것 같네요.

```
new IllegalDelimiterArgumentException(input);

--------

public class IllegalDelimiterArgumentException extends IllegalArgumentException {
    private static final String DEFAULT_MESSAGE = "잘못된 구분자가 입력 되었습니다. 구분자 : %s";

    ...
}
```

추가로, 테스트코드에서 Exception 을 여전히 IllegalArgumentException 으로 검증하고 있네요.
테스트코드도 유지보수가 필요하기때문에 리팩터링할 때 테스트코드도 같이 체크해주세요.

boradol replied:
리뷰어님과 지난번 DM에서 얘기한대로 `IllegalDelimiterArgumentException`에 대한 예외를 검증하게끔 테스트 케이스를 바꿔서 3단계 미션때 적용해보겠습니다 :)

wlroh commented:
> 이 부분에 대해서 수정해 보았는데요. 리뷰어님이 말씀하신대로 제가 잘 이해하고 적용한건지 모르겠더라구요ㅠ
아래 커밋 부분 남깁니다. 확인 부탁드려용~!!

정확히 이해하셨어요! 💯 👍  이미 0을 뜻하는 Number 를 캐싱하셨길래 이왕 만든거 input으로 0이 들어온 케이스도 새로운 객체 생성하는 것보다는 캐싱을 활용하면 좋지않을까 의미였어요

wlroh reviewed (approved):
보라님 안녕하세요 😄 
피드백 잘 반영해주셨습니다 👍 
몇가지 코멘트 남겼는데 확인부탁드려요.
그럼 다음 단계도 화이팅입니다 🔥

boradol commented:
> > 이 부분에 대해서 수정해 보았는데요. 리뷰어님이 말씀하신대로 제가 잘 이해하고 적용한건지 모르겠더라구요ㅠ
> > 아래 커밋 부분 남깁니다. 확인 부탁드려용~!!
> 
> 정확히 이해하셨어요! 💯 👍 이미 0을 뜻하는 Number 를 캐싱하셨길래 이왕 만든거 input으로 0이 들어온 케이스도 새로운 객체 생성하는 것보다는 캐싱을 활용하면 좋지않을까 의미였어요

네네 저도 이 부분 좋았던것 같아요.
코멘트 주신 부분 모두 생각해볼수 있고 참고할 만한 글들도 흥미롭고 재밌어요!
감사합니다. 리뷰어님 짱 👍
